File: clients/android/app/src/main/java/network/buildit/core/sync/OfflineQueue.kt
package network.buildit.core.sync

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Entity
import androidx.room.Index
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Update
import kotlinx.coroutines.flow.Flow
import java.util.UUID

/**
 * Type of operation that can be queued for offline sync.
 */
enum class OperationType {
    MESSAGE,
    FORM_SUBMISSION,
    EVENT_RSVP,
    PROFILE_UPDATE,
    REACTION,
    GROUP_ACTION,
    MUTUAL_AID_REQUEST,
    VOTE,
    WIKI_EDIT,
    CONTACT_NOTE
}

/**
 * Status of a queued operation.
 */
enum class OperationStatus {
    PENDING,
    PROCESSING,
    COMPLETED,
    FAILED,
    CANCELLED
}

/**
 * Room entity for queued operations.
 *
 * Represents an operation that needs to be synchronized when the device is online.
 * Operations are persisted to survive app restarts and are processed with retry logic.
 */
@Entity(
    tableName = "offline_queue",
    indices = [
        Index("status"),
        Index("createdAt"),
        Index("type"),
        Index("priority")
    ]
)
data class QueuedOperationEntity(
    @PrimaryKey
    val id: String = UUID.randomUUID().toString(),

    /** Type of operation */
    val type: OperationType,

    /** JSON-encoded payload data */
    val payload: String,

    /** Optional recipient public key for messages */
    val recipientPublicKey: String? = null,

    /** Optional group ID for group-related operations */
    val groupId: String? = null,

    /** When the operation was created */
    val createdAt: Long = System.currentTimeMillis(),

    /** Current status of the operation */
    val status: OperationStatus = OperationStatus.PENDING,

    /** Number of retry attempts */
    val retryCount: Int = 0,

    /** Last attempt timestamp */
    val lastAttemptAt: Long? = null,

    /** Error message if failed */
    val errorMessage: String? = null,

    /** Priority (higher = more important) */
    val priority: Int = 0
) {
    companion object {
        /** Maximum retry attempts before giving up */
        const val MAX_RETRY_ATTEMPTS = 10

        /** Maximum age for operations (7 days in milliseconds) */
        const val MAX_OPERATION_AGE_MS = 7L * 24 * 60 * 60 * 1000
    }

    /**
     * Calculate the next retry time using exponential backoff.
     * Returns null if the operation shouldn't be retried.
     */
    fun getNextRetryTime(): Long? {
        if (status == OperationStatus.COMPLETED || status == OperationStatus.CANCELLED) {
            return null
        }
        if (retryCount >= MAX_RETRY_ATTEMPTS) {
            return null
        }
        if (retryCount == 0) {
            return System.currentTimeMillis()
        }

        // Exponential backoff: 2^retryCount seconds, capped at 30 minutes
        val delaySeconds = minOf(1 shl retryCount, 1800).toLong()
        val delayMs = delaySeconds * 1000

        return (lastAttemptAt ?: System.currentTimeMillis()) + delayMs
    }

    /**
     * Whether the operation can be retried.
     */
    fun canRetry(): Boolean {
        return retryCount < MAX_RETRY_ATTEMPTS &&
            status != OperationStatus.COMPLETED &&
            status != OperationStatus.CANCELLED
    }

    /**
     * Whether the operation is ready for retry.
     */
    fun isReadyForRetry(): Boolean {
        if (!canRetry()) return false
        val nextRetry = getNextRetryTime() ?: return false
        return nextRetry <= System.currentTimeMillis()
    }
}

/**
 * Data class for queueing operations (without Entity annotations).
 * Used for creating new operations before inserting into the database.
 */
data class QueuedOperation(
    val id: String = UUID.randomUUID().toString(),
    val type: OperationType,
    val payload: String,
    val recipientPublicKey: String? = null,
    val groupId: String? = null,
    val priority: Int = 0
) {
    fun toEntity(): QueuedOperationEntity {
        return QueuedOperationEntity(
            id = id,
            type = type,
            payload = payload,
            recipientPublicKey = recipientPublicKey,
            groupId = groupId,
            priority = priority
        )
    }
}

/**
 * DAO for offline queue operations.
 */
@Dao
interface OfflineQueueDao {

    // ============== Insert Operations ==============

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(operation: QueuedOperationEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(operations: List<QueuedOperationEntity>)

    // ============== Query Operations ==============

    @Query("SELECT * FROM offline_queue ORDER BY priority DESC, createdAt ASC")
    fun getAllOperations(): Flow<List<QueuedOperationEntity>>

    @Query("SELECT * FROM offline_queue WHERE id = :id")
    suspend fun getById(id: String): QueuedOperationEntity?

    @Query("SELECT * FROM offline_queue WHERE status = :status ORDER BY priority DESC, createdAt ASC")
    suspend fun getByStatus(status: OperationStatus): List<QueuedOperationEntity>

    @Query("SELECT * FROM offline_queue WHERE status = :status ORDER BY priority DESC, createdAt ASC")
    fun getByStatusFlow(status: OperationStatus): Flow<List<QueuedOperationEntity>>

    @Query("""
        SELECT * FROM offline_queue
        WHERE (status = 'PENDING' OR (status = 'FAILED' AND retryCount < :maxRetries))
        ORDER BY priority DESC, createdAt ASC
    """)
    suspend fun getPendingOperations(maxRetries: Int = QueuedOperationEntity.MAX_RETRY_ATTEMPTS): List<QueuedOperationEntity>

    @Query("""
        SELECT * FROM offline_queue
        WHERE (status = 'PENDING' OR (status = 'FAILED' AND retryCount < :maxRetries))
        ORDER BY priority DESC, createdAt ASC
    """)
    fun getPendingOperationsFlow(maxRetries: Int = QueuedOperationEntity.MAX_RETRY_ATTEMPTS): Flow<List<QueuedOperationEntity>>

    @Query("SELECT * FROM offline_queue WHERE type = :type ORDER BY priority DESC, createdAt ASC")
    suspend fun getByType(type: OperationType): List<QueuedOperationEntity>

    @Query("SELECT * FROM offline_queue WHERE type = :type ORDER BY priority DESC, createdAt ASC")
    fun getByTypeFlow(type: OperationType): Flow<List<QueuedOperationEntity>>

    @Query("SELECT * FROM offline_queue WHERE recipientPublicKey = :pubkey ORDER BY createdAt ASC")
    suspend fun getForRecipient(pubkey: String): List<QueuedOperationEntity>

    @Query("SELECT * FROM offline_queue WHERE groupId = :groupId ORDER BY createdAt ASC")
    suspend fun getForGroup(groupId: String): List<QueuedOperationEntity>

    // ============== Count Operations ==============

    @Query("""
        SELECT COUNT(*) FROM offline_queue
        WHERE status = 'PENDING' OR (status = 'FAILED' AND retryCount < :maxRetries)
    """)
    fun getPendingCount(maxRetries: Int = QueuedOperationEntity.MAX_RETRY_ATTEMPTS): Flow<Int>

    @Query("SELECT COUNT(*) FROM offline_queue")
    fun getTotalCount(): Flow<Int>

    @Query("SELECT COUNT(*) FROM offline_queue WHERE status = :status")
    fun getCountByStatus(status: OperationStatus): Flow<Int>

    // ============== Update Operations ==============

    @Update
    suspend fun update(operation: QueuedOperationEntity)

    @Query("UPDATE offline_queue SET status = :status WHERE id = :id")
    suspend fun updateStatus(id: String, status: OperationStatus)

    @Query("""
        UPDATE offline_queue
        SET status = :status, lastAttemptAt = :timestamp
        WHERE id = :id
    """)
    suspend fun markProcessing(id: String, status: OperationStatus = OperationStatus.PROCESSING, timestamp: Long = System.currentTimeMillis())

    @Query("""
        UPDATE offline_queue
        SET status = 'COMPLETED'
        WHERE id = :id
    """)
    suspend fun markCompleted(id: String)

    @Query("""
        UPDATE offline_queue
        SET status = 'FAILED',
            retryCount = retryCount + 1,
            lastAttemptAt = :timestamp,
            errorMessage = :errorMessage
        WHERE id = :id
    """)
    suspend fun markFailed(id: String, errorMessage: String, timestamp: Long = System.currentTimeMillis())

    @Query("""
        UPDATE offline_queue
        SET status = 'PENDING'
        WHERE id = :id AND retryCount < :maxRetries
    """)
    suspend fun markForRetry(id: String, maxRetries: Int = QueuedOperationEntity.MAX_RETRY_ATTEMPTS)

    @Query("UPDATE offline_queue SET status = 'CANCELLED' WHERE id = :id")
    suspend fun cancel(id: String)

    // ============== Delete Operations ==============

    @Delete
    suspend fun delete(operation: QueuedOperationEntity)

    @Query("DELETE FROM offline_queue WHERE id = :id")
    suspend fun deleteById(id: String)

    @Query("DELETE FROM offline_queue WHERE status = 'COMPLETED'")
    suspend fun deleteCompleted()

    @Query("DELETE FROM offline_queue WHERE status = 'CANCELLED'")
    suspend fun deleteCancelled()

    @Query("DELETE FROM offline_queue WHERE createdAt < :cutoffTime AND status != 'PROCESSING'")
    suspend fun deleteOlderThan(cutoffTime: Long)

    @Query("DELETE FROM offline_queue")
    suspend fun deleteAll()

    // ============== Cleanup Operations ==============

    /**
     * Prune expired operations (older than 7 days) and completed operations (older than 1 day).
     */
    @Query("""
        DELETE FROM offline_queue
        WHERE (createdAt < :expiredCutoff AND status != 'PROCESSING')
           OR (status = 'COMPLETED' AND createdAt < :completedCutoff)
    """)
    suspend fun pruneExpired(
        expiredCutoff: Long = System.currentTimeMillis() - QueuedOperationEntity.MAX_OPERATION_AGE_MS,
        completedCutoff: Long = System.currentTimeMillis() - (24 * 60 * 60 * 1000)
    )

    /**
     * Reset any stuck processing operations back to pending.
     * This handles cases where the app crashed during processing.
     */
    @Query("""
        UPDATE offline_queue
        SET status = 'PENDING'
        WHERE status = 'PROCESSING' AND lastAttemptAt < :stuckThreshold
    """)
    suspend fun resetStuckOperations(stuckThreshold: Long = System.currentTimeMillis() - (5 * 60 * 1000))
}

File: clients/ios/BuildIt/Core/Sync/OfflineQueue.swift
// OfflineQueue.swift
// BuildIt - Decentralized Mesh Communication
//
// Queue for pending operations (messages, form submissions, etc.)
// Persists queue to disk and handles retry logic with exponential backoff.

import Foundation
import Combine
import os.log

/// Type of operation that can be queued for offline sync
enum OperationType: String, Codable, CaseIterable {
    case message = "message"
    case formSubmission = "form_submission"
    case eventRsvp = "event_rsvp"
    case profileUpdate = "profile_update"
    case reaction = "reaction"
    case groupAction = "group_action"
    case mutualAidRequest = "mutual_aid_request"
    case vote = "vote"
    case wikiEdit = "wiki_edit"
    case contactNote = "contact_note"
}

/// Status of a queued operation
enum OperationStatus: String, Codable {
    case pending = "pending"
    case processing = "processing"
    case completed = "completed"
    case failed = "failed"
    case cancelled = "cancelled"
}

/// Represents a queued operation for offline sync
struct QueuedOperation: Codable, Identifiable, Equatable {
    let id: String
    let type: OperationType
    let payload: Data
    let recipientPublicKey: String?
    let groupId: String?
    let createdAt: Date
    var status: OperationStatus
    var retryCount: Int
    var lastAttemptAt: Date?
    var errorMessage: String?
    var priority: Int

    /// Next retry time based on exponential backoff
    var nextRetryAt: Date? {
        guard status == .failed || status == .pending else { return nil }
        guard retryCount > 0 else { return Date() }

        // Exponential backoff: 2^retryCount seconds, capped at 30 minutes
        let delay = min(pow(2.0, Double(retryCount)), 1800)
        return lastAttemptAt?.addingTimeInterval(delay)
    }

    /// Whether the operation can be retried
    var canRetry: Bool {
        retryCount < QueuedOperation.maxRetryAttempts && status != .completed && status != .cancelled
    }

    /// Maximum retry attempts before giving up
    static let maxRetryAttempts = 10

    init(
        id: String = UUID().uuidString,
        type: OperationType,
        payload: Data,
        recipientPublicKey: String? = nil,
        groupId: String? = nil,
        createdAt: Date = Date(),
        status: OperationStatus = .pending,
        retryCount: Int = 0,
        lastAttemptAt: Date? = nil,
        errorMessage: String? = nil,
        priority: Int = 0
    ) {
        self.id = id
        self.type = type
        self.payload = payload
        self.recipientPublicKey = recipientPublicKey
        self.groupId = groupId
        self.createdAt = createdAt
        self.status = status
        self.retryCount = retryCount
        self.lastAttemptAt = lastAttemptAt
        self.errorMessage = errorMessage
        self.priority = priority
    }

    static func == (lhs: QueuedOperation, rhs: QueuedOperation) -> Bool {
        lhs.id == rhs.id
    }
}

/// OfflineQueue manages a persistent queue of operations for offline sync
actor OfflineQueue {
    // MARK: - Singleton

    static let shared = OfflineQueue()

    // MARK: - Properties

    private var queue: [QueuedOperation] = []
    private let logger = Logger(subsystem: "com.buildit", category: "OfflineQueue")
    private let encryption = DatabaseEncryption.shared
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    /// Publisher for queue updates
    private let queueSubject = PassthroughSubject<[QueuedOperation], Never>()
    nonisolated var queuePublisher: AnyPublisher<[QueuedOperation], Never> {
        queueSubject.eraseToAnyPublisher()
    }

    /// Maximum queue size
    private let maxQueueSize = 500

    /// Maximum age for operations (7 days)
    private let maxOperationAge: TimeInterval = 7 * 24 * 60 * 60

    // MARK: - Initialization

    private init() {
        Task {
            await loadFromDisk()
            await pruneExpiredOperations()
        }
    }

    // MARK: - Queue Operations

    /// Enqueue a new operation
    func enqueue(_ operation: QueuedOperation) {
        // Check for duplicate
        if let existingIndex = queue.firstIndex(where: { $0.id == operation.id }) {
            queue[existingIndex] = operation
            logger.info("Updated existing operation: \(operation.id)")
        } else {
            queue.append(operation)
            logger.info("Enqueued new operation: \(operation.id) type: \(operation.type.rawValue)")
        }

        // Enforce queue size limit
        while queue.count > maxQueueSize {
            // Remove oldest completed/failed operations first
            if let index = queue.firstIndex(where: { $0.status == .completed }) {
                queue.remove(at: index)
            } else if let index = queue.firstIndex(where: { $0.status == .failed && !$0.canRetry }) {
                queue.remove(at: index)
            } else {
                // Remove oldest pending operation
                queue.removeFirst()
            }
        }

        saveToDisk()
        notifyQueueUpdate()
    }

    /// Enqueue a message operation
    func enqueueMessage(content: String, to recipientPublicKey: String, groupId: String? = nil) {
        let payload = try? encoder.encode(["content": content])
        let operation = QueuedOperation(
            type: .message,
            payload: payload ?? Data(),
            recipientPublicKey: recipientPublicKey,
            groupId: groupId,
            priority: 1
        )
        enqueue(operation)
    }

    /// Enqueue a form submission
    func enqueueFormSubmission(formId: String, data: [String: Any]) {
        let payload = try? JSONSerialization.data(withJSONObject: data)
        let operation = QueuedOperation(
            type: .formSubmission,
            payload: payload ?? Data(),
            priority: 0
        )
        enqueue(operation)
    }

    /// Get all pending operations
    func getPendingOperations() -> [QueuedOperation] {
        queue.filter { $0.status == .pending || ($0.status == .failed && $0.canRetry) }
            .sorted { ($0.priority, $0.createdAt) > ($1.priority, $1.createdAt) }
    }

    /// Get operations ready for retry
    func getOperationsReadyForRetry() -> [QueuedOperation] {
        let now = Date()
        return queue.filter { operation in
            guard operation.canRetry else { return false }
            guard let nextRetry = operation.nextRetryAt else { return true }
            return nextRetry <= now
        }.sorted { ($0.priority, $0.createdAt) > ($1.priority, $1.createdAt) }
    }

    /// Get operations by type
    func getOperations(ofType type: OperationType) -> [QueuedOperation] {
        queue.filter { $0.type == type }
    }

    /// Get operation by ID
    func getOperation(id: String) -> QueuedOperation? {
        queue.first { $0.id == id }
    }

    /// Get pending count
    var pendingCount: Int {
        queue.filter { $0.status == .pending || ($0.status == .failed && $0.canRetry) }.count
    }

    /// Get total queue count
    var totalCount: Int {
        queue.count
    }

    // MARK: - Status Updates

    /// Mark an operation as processing
    func markProcessing(_ operationId: String) {
        guard let index = queue.firstIndex(where: { $0.id == operationId }) else { return }
        queue[index].status = .processing
        queue[index].lastAttemptAt = Date()
        saveToDisk()
        notifyQueueUpdate()
        logger.info("Operation processing: \(operationId)")
    }

    /// Mark an operation as completed
    func markCompleted(_ operationId: String) {
        guard let index = queue.firstIndex(where: { $0.id == operationId }) else { return }
        queue[index].status = .completed
        saveToDisk()
        notifyQueueUpdate()
        logger.info("Operation completed: \(operationId)")
    }

    /// Mark an operation as failed
    func markFailed(_ operationId: String, error: String) {
        guard let index = queue.firstIndex(where: { $0.id == operationId }) else { return }
        queue[index].status = .failed
        queue[index].retryCount += 1
        queue[index].lastAttemptAt = Date()
        queue[index].errorMessage = error

        if queue[index].canRetry {
            queue[index].status = .pending
            logger.warning("Operation failed (will retry): \(operationId) - \(error)")
        } else {
            logger.error("Operation permanently failed: \(operationId) - \(error)")
        }

        saveToDisk()
        notifyQueueUpdate()
    }

    /// Cancel an operation
    func cancel(_ operationId: String) {
        guard let index = queue.firstIndex(where: { $0.id == operationId }) else { return }
        queue[index].status = .cancelled
        saveToDisk()
        notifyQueueUpdate()
        logger.info("Operation cancelled: \(operationId)")
    }

    /// Remove completed operations
    func removeCompleted() {
        queue.removeAll { $0.status == .completed }
        saveToDisk()
        notifyQueueUpdate()
    }

    /// Remove an operation from the queue
    func remove(_ operationId: String) {
        queue.removeAll { $0.id == operationId }
        saveToDisk()
        notifyQueueUpdate()
    }

    /// Clear all operations
    func clear() {
        queue.removeAll()
        saveToDisk()
        notifyQueueUpdate()
        logger.info("Queue cleared")
    }

    // MARK: - Batch Operations

    /// Dequeue all pending operations for processing
    func dequeueAllPending() -> [QueuedOperation] {
        let pending = getOperationsReadyForRetry()
        for operation in pending {
            if let index = queue.firstIndex(where: { $0.id == operation.id }) {
                queue[index].status = .processing
                queue[index].lastAttemptAt = Date()
            }
        }
        saveToDisk()
        notifyQueueUpdate()
        return pending
    }

    // MARK: - Persistence

    private func saveToDisk() {
        guard let data = try? encoder.encode(queue) else {
            logger.error("Failed to encode queue")
            return
        }

        let url = getQueueFileURL()
        do {
            try encryption.writeEncrypted(data, to: url)
        } catch {
            logger.error("Failed to save queue: \(error.localizedDescription)")
        }
    }

    private func loadFromDisk() {
        let url = getQueueFileURL()
        let legacyUrl = getDocumentsDirectory().appendingPathComponent("offline_queue.json")

        do {
            if let data = try encryption.readEncrypted(from: url),
               let loaded = try? decoder.decode([QueuedOperation].self, from: data) {
                queue = loaded
                logger.info("Loaded \(loaded.count) operations from encrypted queue")
            } else if FileManager.default.fileExists(atPath: legacyUrl.path),
                      let data = try? Data(contentsOf: legacyUrl),
                      let loaded = try? decoder.decode([QueuedOperation].self, from: data) {
                queue = loaded
                logger.info("Migrated \(loaded.count) operations from legacy queue")
                try? FileManager.default.removeItem(at: legacyUrl)
                saveToDisk() // Save to encrypted format
            }
        } catch {
            logger.error("Failed to load queue: \(error.localizedDescription)")
        }

        notifyQueueUpdate()
    }

    private func pruneExpiredOperations() {
        let cutoff = Date().addingTimeInterval(-maxOperationAge)
        let initialCount = queue.count

        queue.removeAll { operation in
            // Remove expired pending/failed operations
            if operation.createdAt < cutoff && operation.status != .processing {
                return true
            }
            // Remove completed operations older than 1 day
            if operation.status == .completed && operation.createdAt < Date().addingTimeInterval(-86400) {
                return true
            }
            return false
        }

        let removedCount = initialCount - queue.count
        if removedCount > 0 {
            logger.info("Pruned \(removedCount) expired operations")
            saveToDisk()
            notifyQueueUpdate()
        }
    }

    private func getQueueFileURL() -> URL {
        getDocumentsDirectory().appendingPathComponent("offline_queue.db")
    }

    private func getDocumentsDirectory() -> URL {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
    }

    private func notifyQueueUpdate() {
        let currentQueue = queue
        Task { @MainActor in
            queueSubject.send(currentQueue)
        }
    }
}

// MARK: - Notification Names

extension Notification.Name {
    static let offlineQueueUpdated = Notification.Name("offlineQueueUpdated")
    static let operationCompleted = Notification.Name("operationCompleted")
    static let operationFailed = Notification.Name("operationFailed")
}
